nixos/modules/programs/starship.nix

```diff
@@ -0,0 +1,62 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+
+  cfg = config.programs.starship;
+
+in {
+  meta.maintainers = [ maintainers.imalison ];
+
+  options.programs.starship = {
+    enable = mkEnableOption "starship";
+
+    package = mkOption {
+      type = types.package;
+      default = pkgs.starship;
+      defaultText = literalExample "pkgs.starship";
+      description = "The package to use for the starship binary.";
+    };
+
+    enableBashIntegration = mkOption {
+      default = true;
+      type = types.bool;
+      description = ''
+        Whether to enable Bash integration.
+      '';
+    };
+
+    enableZshIntegration = mkOption {
+      default = true;
+      type = types.bool;
+      description = ''
+        Whether to enable Zsh integration.
+      '';
+    };
+
+    enableFishIntegration = mkOption {
+      default = true;
+      type = types.bool;
+      description = ''
+        Whether to enable Fish integration.
+      '';
+    };
+  };
+
+  config = mkIf cfg.enable {
+    environment.systemPackages = [ cfg.package ];
+
+    programs.bash.promptInit = mkIf cfg.enableBashIntegration ''
+      if [[ $TERM != "dumb" && (-z $INSIDE_EMACS || $INSIDE_EMACS == "vterm" || ''${STARSHIP_INSIDE_EMACS-no} != "no") ]]; then
```

Comment:
How is this `STARSHIP_INSIDE_EMACS-no` variable defined? `home-manager` doesn't have it.

Reply:
Also can we have it configurable for midnight commander? `starship` is not very useful there and can slow down mc. It can be checked via `MC_SID` variable like this: https://github.com/kurnevsky/dotfiles/blob/92ddfbbe53ee8cfcef5de7d9f2e58cf4db930c93/.zshrc#L161

Reply:
> How is this `STARSHIP_INSIDE_EMACS` variable defined? Or what is it used for? `home-manager` doesn't have it.

It's not defined by default, however you are free to add something I itializing it, which would then allow the use of starship inside of emacs. This must be done before this initialization but it is certainly possible.

I considered using a nix option to control this, but I actually want to be able to dynamically control this, and this approach also let's different users do different things.

Reply:
> Also can we have it configurable for midnight commander? `starship` is not very useful there and can slow down mc. It can be checked via `MC_SID` variable like this: https://github.com/kurnevsky/dotfiles/blob/92ddfbbe53ee8cfcef5de7d9f2e58cf4db930c93/.zshrc#L161

Hmm sure, but this does make me think that somehow we are engaging in a bit of an anti-pattern by having logic for specific programs in here. Perhaps a better approach would be to just always use a variable like starship-inside-emacs, and allow users to manually control things with further config that way?

Reply:
> Perhaps a better approach would be to just always use a variable like starship-inside-emacs, and allow users to manually control things with further config that way?

Yeah, it might be better. But then we need to make/rename this variable to something more generic and have to make sure that users are able to insert their code before this init is executed. And I'd leave this emacs machinery anyway, because it's a good default.

Reply:
>  to make sure that users are able to insert their code before this init is executed. 

Probably just create a custom `initBefore` variable for starship.